internal/controller/dashboard_controller.go

```diff
@@ -0,0 +1,344 @@
+				var dashboards v1.ConfigMapList
+
+				err := mgr.GetClient().List(ctx, &dashboards, client.MatchingLabels{"app.giantswarm.io/kind": "dashboard"})
+				if err != nil {
+					logger.Error(err, "failed to list grafana dashboard configmaps")
+					return []reconcile.Request{}
+				}
+
+				// Reconcile all grafana dashboards when the grafana pod is recreated
+				requests := make([]reconcile.Request, 0, len(dashboards.Items))
+				for _, dashboard := range dashboards.Items {
+					requests = append(requests, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Name: dashboard.Name,
+						},
+					})
+				}
+				return requests
+			}),
+			builder.WithPredicates(predicates.GrafanaPodRecreatedPredicate{}),
+		).
+		Complete(r)
+}
+
+// reconcileCreate creates the dashboard.
+// reconcileCreate ensures the Grafana dashboard described in configmap is created in Grafana.
+// This function is also responsible for:
+// - Adding the finalizer to the configmap
+func (r DashboardReconciler) reconcileCreate(ctx context.Context, dashboard *v1.ConfigMap) (ctrl.Result, error) { // nolint:unparam
+	logger := log.FromContext(ctx)
+
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	if !controllerutil.ContainsFinalizer(dashboard, DashboardFinalizer) {
+		// We use a patch rather than an update to avoid conflicts when multiple controllers are adding their finalizer to the grafana dashboard
+		// We use the patch from sigs.k8s.io/cluster-api/util/patch to handle the patching without conflicts
+		logger.Info("adding finalizer", "finalizer", DashboardFinalizer)
+		patchHelper, err := patch.NewHelper(dashboard, r.Client)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+		controllerutil.AddFinalizer(dashboard, DashboardFinalizer)
+		if err := patchHelper.Patch(ctx, dashboard); err != nil {
+			logger.Error(err, "failed to add finalizer", "finalizer", DashboardFinalizer)
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+		logger.Info("added finalizer", "finalizer", DashboardFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Configure the dashboard in Grafana
+	if err := r.configureDashboard(ctx, dashboard); err != nil {
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func getDashboardUID(jsonDashboard string) (UID string, err error) {
+
+	var Dashboard map[string]interface{}
+	err = json.Unmarshal([]byte(jsonDashboard), &Dashboard)
+	if err != nil {
+		return "", err
+	}
+
+	UID, ok := Dashboard["uid"].(string)
+	if !ok {
+		return "", errors.New("dashboard UID not found in configmap")
+	}
+	return UID, nil
+}
+
+func getDashboardCMOrg(dashboard *v1.ConfigMap) (Org string, err error) {
+
+	// Try to look for an annotation first
+	orgName := dashboard.GetAnnotations()[DashboardTeamAnnotation]
+	if orgName != "" {
+		return orgName, nil
+	}
+
+	// Then look for a label
+	orgName = dashboard.GetLabels()[DashboardTeamAnnotation]
```

Comment:
Please check for return values from GetAnnotations and GetLabels, as it migh be nil

Reply:
I meant that `GetAnnotations` and `GetLabels` both return a map, and map can be `nil` so before accessing the value we should check if the map is not nil otherwise this would cause a panic.

Something like this
```go
	labels := dashboard.GetLabels()
	if labels != nil && labels[organization.OrganizationLabel] != "" {
		return labels[organization.OrganizationLabel], nil
	}
```

Reply:
oh, I thought the "ok" check was preventing this to happen.
All right, I'll do it as you recommend! Looks more explicit!